contracts/ConnectivityTrackerHbbft.sol

```diff
@@ -0,0 +1,243 @@
+    event SetEarlyEpochEndToleranceLevel(uint256 _level);
+    event ReportMissingConnectivity(
+        address indexed reporter,
+        address indexed validator,
+        uint256 indexed blockNumber,
+        uint256 blockHash
+    );
+
+    event ReportReconnect(
+        address indexed reporter,
+        address indexed validator,
+        uint256 indexed blockNumber,
+        uint256 blockHash
+    );
+
+    error AlreadyReported(address reporter, address validator);
+    error CannotReportByFlaggedValidator(address reporter);
+    error InvalidAddress();
+    error InvalidBlockNumber();
+    error OnlyValidator();
+    error ReportTooEarly();
+    error UnknownReconnectReporter(address reporter, address validator);
+
+    modifier onlyValidator() {
+        if (!validatorSetContract.isValidator(msg.sender)) {
+            revert OnlyValidator();
+        }
+        _;
+    }
+
+    modifier onlyValidBlock(uint256 blockNumber) {
+        if (blockNumber > block.number) {
+            revert InvalidBlockNumber();
+        }
+
+        uint256 epochStartBlock = stakingContract.stakingEpochStartBlock();
+        if (block.number < epochStartBlock + minReportAgeBlocks) {
+            revert ReportTooEarly();
+        }
+        _;
+    }
+
+    modifier nonFlagged() {
+        uint256 epoch = currentEpoch();
+
+        if (_flaggedValidators[epoch].contains(msg.sender)) {
+            revert CannotReportByFlaggedValidator(msg.sender);
+        }
+        _;
+    }
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        // Prevents initialization of implementation contract
+        _disableInitializers();
+    }
+
+    function initialize(
+        address _contractOwner,
+        address _validatorSetContract,
+        address _stakingContract,
+        uint256 _minReportAgeBlocks
+    ) external initializer {
+        if (
+            _contractOwner == address(0) ||
+            _validatorSetContract == address(0) ||
+            _stakingContract == address(0)
+        ) {
+            revert InvalidAddress();
+        }
+
+        __Ownable_init();
+        _transferOwnership(_contractOwner);
+
+        validatorSetContract = IValidatorSetHbbft(_validatorSetContract);
+        stakingContract = IStakingHbbft(_stakingContract);
+
+        minReportAgeBlocks = _minReportAgeBlocks;
+        earlyEpochEndToleranceLevel = 2;
+    }
+
+    function setMinReportAge(uint256 _minReportAge) external onlyOwner {
+        minReportAgeBlocks = _minReportAge;
+
+        emit SetMinReportAgeBlocks(_minReportAge);
+    }
+
+    function setEarlyEpochEndToleranceLevel(uint256 _level) external onlyOwner {
+        earlyEpochEndToleranceLevel = _level;
+
+        _decideEarlyEpochEndNeeded(currentEpoch());
+
+        emit SetEarlyEpochEndToleranceLevel(_level);
+    }
+
+    function reportMissingConnectivity(
+        address validator,
+        uint256 blockNumber,
+        uint256 blockHash
```

Comment:
The idea of the blockHash was to apply a minimal protection, and check if the reporting Node is on the correct block.